src/protocols/bdx/tests/TestBdxTransferSession.cpp

```diff
@@ -113,183 +154,490 @@ void VerifyStatusReport(nlTestSuite * inSuite, void * inContext, const System::P
+
+        // Test Ack -> Query -> Block
+        bool isEof = (kNumBlocksSent == kNumBlockSends - 1);
+        kNumBlocksSent++;
+        sResponderExpectedOutputEvent.blockdata.Data = nullptr;
+        SendAndVerifyBlockAck(inSuite, *sBdxReceiver, *sBdxSender, isEof);
+        return true;
+    }
+    case TransferSession::OutputEventType::kInitReceived:
+    case TransferSession::OutputEventType::kStatusReceived:
+    case TransferSession::OutputEventType::kAckEOFReceived:
+    case TransferSession::OutputEventType::kInternalError:
+    case TransferSession::OutputEventType::kTransferTimeout:
+    case TransferSession::OutputEventType::kQueryWithSkipReceived:
+    case TransferSession::OutputEventType::kQueryReceived:
+    case TransferSession::OutputEventType::kNone:
+    case TransferSession::OutputEventType::kAckReceived:
+    default:
+        return false;
     }
 }
 
+// Helper method for initializing two TransferSession objects, generating a TransferInit message, and passing it to a responding
+// TransferSession.
+void SendAndVerifyTransferInit(nlTestSuite * inSuite, TransferSession & initiator, TransferRole initiatorRole,
+                               TransferSession::TransferInitData initData, TransferSession & responder,
+                               BitFlags<TransferControlFlags> & responderControlOpts, uint16_t responderMaxBlock)
+{
+    TransferRole responderRole = (initiatorRole == TransferRole::kSender) ? TransferRole::kReceiver : TransferRole::kSender;
+
+    MessageType expectedInitMsg = (initiatorRole == TransferRole::kSender) ? MessageType::SendInit : MessageType::ReceiveInit;
+
+    // Initializer responder to wait for transfer
+    CHIP_ERROR err =
+        responder.WaitForTransfer(responderRole, responderControlOpts, responderMaxBlock, OnResponderOutputEventReceived, inSuite);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Set the expected output event when init message is sent to the initiator
+    sInitiatorExpectedOutputEvent.EventType               = TransferSession::OutputEventType::kMsgToSend;
+    sInitiatorExpectedOutputEvent.msgTypeData.ProtocolId  = Protocols::BDX::Id;
+    sInitiatorExpectedOutputEvent.msgTypeData.MessageType = static_cast<uint8_t>(expectedInitMsg);
+
+    // Set the expected BDX Init message to the responder
+    sResponderExpectedOutputEvent.EventType        = TransferSession::OutputEventType::kInitReceived;
+    sResponderExpectedOutputEvent.transferInitData = initData;
+
+    // Verify initiator outputs respective Init message (depending on role) after StartTransfer()
+    err = initiator.StartTransfer(initiatorRole, initData, OnInitiatorOutputEventReceived, inSuite);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Now process output events until we get to one that needs no action from
+    // us.
+    bool transferInProgress = true;
+    do
+    {
+        transferInProgress = RespondToLastInitiatorOutputEvent(inSuite);
+    } while (transferInProgress);
```

Comment:
This entire unit test now contains practically more logic than the protocol itself, and could now contain bugs. It would have been great to avoid adding so much actual logic to the unit test.